### PR TITLE
Connect support for dynamic linking to user options

### PR DESCRIPTION
**Pull request by LU-JOHN**
Add option "-fsycl-allow-device-dependencies" to enable support for dynamic linking.

Also:
1. No functions are importable without "-fsycl-allow-device-dependencies"
2. Deal with SYCL_EXTERNAL header decls that have lost SYCL_EXTERNAL attribute in LLVM IR
3. SPIRV/SYCL/ESIMD builtins cannot be an imported function

Tested in three E2E test cases.

Minor change:
Change SYCL-EXTERNAL to SYCL_EXTERNAL in testcase comment.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -4188,6 +4188,10 @@ def fsycl_remove_unused_external_funcs : Flag<["-"], "fsycl-remove-unused-extern
   Group<sycl_Group>, HelpText<"Allow removal of unused `SYCL_EXTERNAL` functions (default)">;
 def fno_sycl_remove_unused_external_funcs : Flag<["-"], "fno-sycl-remove-unused-external-funcs">,
   Group<sycl_Group>, HelpText<"Prevent removal of unused `SYCL_EXTERNAL` functions">;
+def fsycl_allow_split_with_dependencies : Flag<["-"], "fsycl-allow-split-with-dependencies">,
+  Group<sycl_Group>, HelpText<"Allow dependencies between device code images">;
+def fno_sycl_allow_split_with_dependencies : Flag<["-"], "fno-sycl-allow-split-with-dependencies">,
```
Comment (sarnex) [inline clang/include/clang/Driver/Options.td:4193]: I'm not sure if we need approval from the options group for these, @mdtoguchi?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -10712,6 +10712,17 @@ static void addArgs(ArgStringList &DstArgs, const llvm::opt::ArgList &Alloc,
   }
 }
 
+static bool supportDynamicLinking(const llvm::opt::ArgList &TCArgs) {
+  if (TCArgs.hasFlag(options::OPT_fsycl_allow_split_with_dependencies,
```
Comment (sarnex) [inline clang/lib/Driver/ToolChains/Clang.cpp:10716]: I would just inline this into the callsites and move this into a seperate function when the SYCL RT is updated, checking-in commented out code is not ideal IMO

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp [hunk 3]
```diff
@@ -1244,7 +1244,11 @@ splitSYCLModule(std::unique_ptr<Module> M, ModuleSplitterSettings Settings) {
 
 bool canBeImportedFunction(const Function &F) {
   if (F.isIntrinsic() || F.getName().starts_with("__") ||
-      !llvm::sycl::utils::isSYCLExternalFunction(&F))
+      !( // SYCL_EXTERNAL property is not recorded for a declaration
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:1247]: I would put this comment above the `if` instead of in the middle of the condition

Comment (LU-JOHN) [reply llvm/lib/SYCLLowerIR/ModuleSplitter.cpp]: Comment moved.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (LU-JOHN) [reply clang/lib/Driver/ToolChains/Clang.cpp:10716]: Removed commented code.  Changed into TODO comment.

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (mdtoguchi) [reply clang/include/clang/Driver/Options.td:4193]: @sarnex and @LU-JOHN, any new option (that we are inventing) that is meant to be visible should go through the options team workgroup.

Comment (LU-JOHN) [reply clang/include/clang/Driver/Options.td]: Add new option in separate PR.

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp [hunk 4]
```diff
@@ -1243,8 +1243,12 @@ splitSYCLModule(std::unique_ptr<Module> M, ModuleSplitterSettings Settings) {
 }
 
 bool canBeImportedFunction(const Function &F) {
+  // SYCL_EXTERNAL property is not recorded for a declaration
+  // in a header file.  Thus SYCL IR that is a declaration
+  // will be considered as SYCL_EXTERNAL for the purposes of
+  // this function.
   if (F.isIntrinsic() || F.getName().starts_with("__") ||
-      !llvm::sycl::utils::isSYCLExternalFunction(&F))
+      !(F.isDeclaration() || llvm::sycl::utils::isSYCLExternalFunction(&F)))
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:1251]: nit: can we distribute the `!`, so like

`(!F.isDeclaration() && !llvm::sycl::utils::isSYCLExternalFunction(&F))`

it's easier to read IMO

**Review by sarnex — APPROVED** [review]
lgtm only minor comments, the `supportDynamicLinking` function is a bit strange but i'm assuming this is an incremental change and it will be fleshed out soon, so it's fine with me

**Review by mdtoguchi — APPROVED** [review]

File: llvm/lib/SYCLLowerIR/ModuleSplitter.cpp (hunk 4 above)
Comment (LU-JOHN) [reply llvm/lib/SYCLLowerIR/ModuleSplitter.cpp]: Updated expression.

**Review by sarnex — APPROVED** [review]

**Comment by LU-JOHN** [comment]
> lgtm only minor comments, the `supportDynamicLinking` function is a bit strange but i'm assuming this is an incremental change and it will be fleshed out soon, so it's fine with me

Yes, this function can only return true with the -shared option when the SYCL RT is ready.  Otherwise, "-shared" tests may start failing.

**Review by maarquitos14 — APPROVED** [review]
LGTM.

**Comment by AlexeySachkov** [comment]
All builds have successfully completed after merge with `sycl` branch. Considering that we had a fully green CI run on the second to last commit and that changes this PR introduces are into a feature that is not supported on CUDA and HIP, I think that we can merge this PR sooner without waiting for CUDA & HIP CI jobs to complete.

I will proceed with merge, but @LU-JOHN, @maarquitos14, please keep an eye on unfinished jobs in case they highlight any unexpected failures.

**Pull request merged**